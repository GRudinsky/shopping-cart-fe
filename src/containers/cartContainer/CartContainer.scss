@import '../../styles/variables';
@import '../../index.scss';

.cartWrapper {
  width: 450px;
  height: auto;
  display: flex;
  flex-direction: column;
  justify-content: center;
  align-items: center;
  border: 1px solid $color-border;
  border-radius: 3px;
  box-shadow: 1px 1px $color-border;
  background-color: $color-background-container-primary;

  @media (max-width: 640px) {
    width: 100%;
    height: 100%;
  }

  .productWrapper {
    width: 90%;
    display: flex;
    flex-direction: column;
    justify-content: center;
    align-items: center;
    margin: 20px 0px;

    .productRow {
      width: 100%;
      height: 60px;
      line-height: 60px;
      text-align: center;
      display: flex;
      justify-content: space-between;
      align-content: center;

      &:not(:first-of-type) {
        border-top: 1px dashed $color-border;
      }

      .productTitle {
        color: $color-slate;
        display: block;
        text-align: start;
        flex-basis: 48%;
        font-size: $font-medium;
        font-weight: bold;
      }
      .productNumbers {
        display: flex;
        justify-content: space-between;
        vertical-align: middle;
        flex-basis: 52%;

        input {
          margin-bottom: auto;
          margin-top: auto;
          max-width: 20px;
          max-height: 20px;
          text-align: center;
          cursor: pointer;
          border: 1px solid $color-border;
          border-radius: 3px;
          font-size: $font-medium;
          font-weight: bold;
          color: $color-slate;

          &:focus {
            outline-color: $color-info;
          }
        }
        input::-webkit-outer-spin-button,
        input::-webkit-inner-spin-button {
          -webkit-appearance: none;
          margin: 0;
        }

        input[type='number'] {
          -moz-appearance: textfield; /* Firefox */
        }

        .singleProductTotals {
          color: $color-orange;
          font-size: $font-medium;
          font-weight: bold;
        }
      }
    }
  }
  .actionWrapper {
    display: flex;
    width: 100%;
    height: 10vh;
    padding: 10px 0;
    background-color: $color-background-container-secondary;
    border-top: 1px solid $color-border;

    .cartSubTotal {
      color: $color-slate;
      display: block;
      text-align: start;
      flex-basis: 48%;
      margin-left: 20px;
      text-justify: center;
      margin-top: auto;
      margin-bottom: auto;
    }

    .buttonWrapper {
      display: flex;
      justify-content: space-between;
      vertical-align: middle;
      flex-basis: 52%;
      margin-right: 20px;

      button {
        margin-top: auto;
        margin-bottom: auto;
        cursor: pointer;
        font-family: $body-font;
        font-weight: bold;
        color: $color-slate;
        font-size: $font-medium;
      }

      .buttonClear {
        border: none;
        background: transparent;
      }

      .buttonCheckout {
        border: none;
        background-color: $color-info;
        color: $color-background-container-primary;
        width: 60%;
        height: 50px;
      }
    }
  }
}
