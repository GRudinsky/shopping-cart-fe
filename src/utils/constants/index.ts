import { CART_API_URL } from './urls';
import { ERROR_MESSAGE_LIST } from './messages';

export { ERROR_MESSAGE_LIST, CART_API_URL };
