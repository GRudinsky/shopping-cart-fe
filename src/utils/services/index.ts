import { getShoppingCart } from './services';
export { getShoppingCart };
