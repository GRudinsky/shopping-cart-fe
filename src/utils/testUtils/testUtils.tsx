import React from 'react';
import { render, screen, waitFor } from '@testing-library/react';
import { Provider } from 'react-redux';
import store from '../../redux/store';

const renderWithStore = (children: React.ReactElement) => {
  return render(<Provider store={store}>{children}</Provider>);
};

export { renderWithStore, screen, waitFor };
