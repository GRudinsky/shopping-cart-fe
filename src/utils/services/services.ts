import 'whatwg-fetch';
import { CART_API_URL } from '../constants';
export const getShoppingCart = async () => {
  const options = {
    method: 'GET'
  };
  try {
    const response = await fetch(CART_API_URL, options);
    return await response.json();
  } catch (err: any) {
    throw new Error(err.message);
  }
};
