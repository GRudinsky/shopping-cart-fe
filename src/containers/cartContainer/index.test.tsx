import userEvent from '@testing-library/user-event';
import { renderWithStore, screen, waitFor } from '../../utils/testUtils';
import { getShoppingCart } from '../../utils/services';
import { shoppingCartResponse } from '../../utils/fixtures';
import CartContainer from '.';

jest.mock('../../utils/services');

beforeEach(() => {
  (getShoppingCart as jest.Mock).mockResolvedValue(shoppingCartResponse);
});

afterEach(() => {
  (getShoppingCart as jest.Mock).mockReset();
});

describe('<CartContainer/>', () => {
  it('should not render the cart container when getShoppingCart returns null', () => {
    (getShoppingCart as jest.Mock).mockResolvedValue(null);
    const { container } = renderWithStore(<CartContainer />);

    expect(screen.queryByTestId('cartWrapper')).not.toBeInTheDocument();
    expect(container).toMatchSnapshot();
  });

  it('should render the cart container correctly when getShoppingCart call returns data', async () => {
    const { container } = renderWithStore(<CartContainer />);

    expect(await screen.findAllByRole('button')).toHaveLength(2);
    expect(screen.getAllByTestId('productRow')).toHaveLength(3);
    expect(container).toMatchSnapshot();
  });

  it('should update the total product and shopping cart values when the product quantity changes', async () => {
    renderWithStore(<CartContainer />);

    const input = await screen.findByTestId('quantityInput_0');
    const totalProductValue = screen.getByTestId('productTotals_0');
    const totalCartValue = screen.getByTestId('cartSubTotal');

    expect(input).toHaveValue(1);
    expect(totalProductValue).toHaveTextContent('$1.80');
    expect(totalCartValue).toHaveTextContent('$8.25');

    userEvent.tab();
    expect(input).toHaveFocus();

    userEvent.clear(input);
    userEvent.type(input, '3');

    await waitFor(() => {
      expect(input).toHaveValue(3);
    });

    expect(totalProductValue).toHaveTextContent('$5.40');
    expect(totalCartValue).toHaveTextContent('$11.85');
  });

  it('should reset the total product and shopping cart values when the "Clear" button is clicked', async () => {
    renderWithStore(<CartContainer />);

    const input = await screen.findByTestId('quantityInput_0');
    const totalProductValue1 = screen.getByTestId('productTotals_0');
    const totalProductValue2 = screen.getByTestId('productTotals_1');
    const totalProductValue3 = screen.getByTestId('productTotals_2');
    const totalCartValue = screen.getByTestId('cartSubTotal');
    const button = screen.getByTestId('buttonClear');

    expect(input).toHaveValue(3);
    expect(totalProductValue1).toHaveTextContent('$5.40');
    expect(totalProductValue2).toHaveTextContent('$3.10');
    expect(totalProductValue3).toHaveTextContent('$3.35');
    expect(totalCartValue).toHaveTextContent('$11.85');

    userEvent.click(button);

    await waitFor(() => {
      expect(input).toHaveValue(0);
    });
    expect(totalProductValue1).toHaveTextContent('$0');
    expect(totalProductValue2).toHaveTextContent('$0');
    expect(totalProductValue3).toHaveTextContent('$0');
    expect(totalCartValue).toHaveTextContent('$0');
  });
});
