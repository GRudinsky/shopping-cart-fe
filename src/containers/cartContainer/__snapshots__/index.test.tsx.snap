// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<CartContainer/> should not render the cart container when getShoppingCart returns null 1`] = `<div />`;

exports[`<CartContainer/> should render the cart container correctly when getShoppingCart call returns data 1`] = `
<div>
  <div
    class="cartWrapper"
    id="cartWrapper"
  >
    <div
      class="productWrapper"
    >
      <div
        class="productRow"
        data-testid="productRow"
      >
        <h6
          class="productTitle"
          data-testid="productTitle_0"
        >
          Mountain Dew
        </h6>
        <div
          class="productNumbers"
        >
          <input
            data-testid="quantityInput_0"
            max="99"
            min="0"
            name="1"
            type="number"
            value="1"
          />
          <h6
            class="singleProductTotals"
            data-testid="productTotals_0"
          >
            $1.80
          </h6>
        </div>
      </div>
      <div
        class="productRow"
        data-testid="productRow"
      >
        <h6
          class="productTitle"
          data-testid="productTitle_1"
        >
          Desperados
        </h6>
        <div
          class="productNumbers"
        >
          <input
            data-testid="quantityInput_1"
            max="99"
            min="0"
            name="2"
            type="number"
            value="1"
          />
          <h6
            class="singleProductTotals"
            data-testid="productTotals_1"
          >
            $3.10
          </h6>
        </div>
      </div>
      <div
        class="productRow"
        data-testid="productRow"
      >
        <h6
          class="productTitle"
          data-testid="productTitle_2"
        >
          Jack Daniels
        </h6>
        <div
          class="productNumbers"
        >
          <input
            data-testid="quantityInput_2"
            max="99"
            min="0"
            name="3"
            type="number"
            value="1"
          />
          <h6
            class="singleProductTotals"
            data-testid="productTotals_2"
          >
            $3.35
          </h6>
        </div>
      </div>
    </div>
    <div
      class="actionWrapper"
    >
      <h2
        class="cartSubTotal"
        data-testid="cartSubTotal"
      >
        $8.25
      </h2>
      <div
        class="buttonWrapper"
      >
        <button
          class="buttonClear"
          data-testid="buttonClear"
        >
          Clear
        </button>
        <button
          class="buttonCheckout"
          data-testid="buttonCheckout"
        >
          Check Out
        </button>
      </div>
    </div>
  </div>
</div>
`;
